photon-ml/src/main/scala/com/linkedin/photon/ml/data/ScoredGameDatum.scala

```diff
@@ -0,0 +1,118 @@
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.data
+
+import scala.collection.Map
+
+import com.linkedin.photon.ml.util.Utils.isAlmostZero
+
+/**
+  * Representation of a single scored GAME data point
+  *
+  * @param response The response or label
+  * @param offset offset
+  * @param weight importance weight
+  * @param score computed score for this instance
+  * @param idTypeToValueMap The id type to value map that holds different types of ids associated with this data
+  *                         point. A few examples of the ids types are: (i) ids used to build the random effect model
+  *                         such as userId and itemId; (ii) ids used to compute certain metrics like precision@k such
+  *                         as documentId or queryId; (iii) ids that are used to uniquely identify each training record.
+  */
+case class ScoredGameDatum(response: Double, offset: Double, weight: Double,
+                           var score: Double, idTypeToValueMap: Map[String, String]) extends Serializable {
+  /**
+    * Get a copy of the current instance with a score of [[ScoredGameDatum.DEFAULT_SCORE]]
+    *
+    * @return copy of the current scored datum instance with zero score
+    */
+  def getZeroScoreDatum(): ScoredGameDatum = {
+    this.copy(score = ScoredGameDatum.DEFAULT_SCORE)
+  }
+
+  /**
+    * Two scored data are defined to be similar if they have identical [[response]], [[offset]], [[weight]] and
+    * [[idTypeToValueMap]]. Their scores are allowed to be different
+    *
+    * @param that the other scored datum instance
+    * @return boolean indicating whether the data are similar
+    */
+  def similarTo(that: ScoredGameDatum): Boolean = {
+    ((this.response.isNaN && that.response.isNaN) || isAlmostZero(this.response - that.response)) &&
+      isAlmostZero(this.offset - that.offset) &&
+      isAlmostZero(this.weight - that.weight) &&
+      this.idTypeToValueMap.equals(that.idTypeToValueMap)
+  }
+
+  /**
+    * The plus operation for scored data. Two data can only be added if their [[response]], [[offset]], [[weight]] and
+    * [[idTypeToValueMap]] are identical. See [[similarTo()]]
+    *
+    * @param that The other scored datum instance
+    * @return A new scored datum instance with score being the sum of the two input scores
+    */
+  def +(that: ScoredGameDatum): ScoredGameDatum = {
+    require(similarTo(that), s"Can only add two ScoredGameDatum objects if their response, offset, weight" +
```

Comment:
Also, adding a `score` field to `GameDatum` directly could further simplify, unless there is a solid reason to have a new `ScoredGameDatum` class that I missed?